src/test/app/AMM_test.cpp

```diff
@@ -4132,6 +4132,341 @@ struct AMM_test : public jtx::AMMTest
+                                STAmount{ETH, UINT64_C(64'91106406735152), -14},
+                            }}}));
+                    }
+                    else
+                    {
+                        // Ed offer is partially crossed.
+                        BEAST_EXPECT(expectOffers(
+                            env,
+                            ed,
+                            1,
+                            {{Amounts{
+                                STAmount{ETH, UINT64_C(335'0889359326485), -13},
+                                STAmount{USD, UINT64_C(209'4305849579053), -13},
+                            }}}));
+                        BEAST_EXPECT(expectOffers(env, alice, 0));
+                    }
+                }
+            }
+
+            // Strand selection
+
+            // Two book steps strand quality is 1.
+            // AMM strand's best quality is equal to AMM's spot price
+            // quality, which is 1. Both strands (steps) are adjusted
+            // for the transfer fee in qualityUpperBound. In case
+            // of two strands, AMM offers have better quality and are consumed
+            // first, remaining liquidity is generated by CLOB offers.
+            // Liquidity from two strands is better in this case than in case
+            // of one strand with two book steps. Liquidity from one strand
+            // with AMM has better quality than either one strand with two book
+            // steps or two strands. It may appear unintuitive, but one strand
+            // with AMM is optimized and generates one AMM offer, while in case
+            // of two strands, multiple AMM offers are generated, which results
+            // in slightly worse overall quality.
+            {
+                std::array<Quality, 3> q;
+                for (auto i = 0; i < 3; ++i)
+                {
+                    Env env(*this);
+                    prep(env, rates.first, rates.second);
+                    std::shared_ptr<AMM> amm;
+
+                    if (i == 0 || i == 2)
+                    {
+                        env(offer(ed, ETH(400), CAN(400)), txflags(tfPassive));
+                        env(offer(ed, CAN(400), USD(400))), txflags(tfPassive);
+                        env.close();
+                    }
+
+                    if (i > 0)
+                        amm = std::make_shared<AMM>(
+                            env, ed, ETH(1'000), USD(1'000));
+
+                    env(pay(carol, bob, USD(100)),
+                        path(~USD),
+                        path(~CAN, ~USD),
+                        sendmax(ETH(600)),
+                        txflags(tfPartialPayment));
+                    env.close();
+
+                    BEAST_EXPECT(expectLine(env, bob, USD(2'100)));
+
+                    if (i == 2)
+                    {
+                        // Liquidity is consumed from AMM strand only
+                        if (rates.first == 1.5)
+                        {
```

Comment:
This comment only applies to the `rates.first == 1.5`, correct? If so, then it should be inside that block.